common/autoware_agnocast_wrapper/include/autoware/agnocast_wrapper/autoware_agnocast_wrapper.hpp

```diff
@@ -0,0 +1,67 @@
+// Copyright 2025 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#ifdef USE_AGNOCAST_ENABLED
+
+#include <agnocast/agnocast.hpp>
+
+#define AUTOWARE_MESSAGE_PTR(MessageT) agnocast::ipc_shared_ptr<MessageT>
+#define AUTOWARE_SUBSCRIPTION_PTR(MessageT) typename agnocast::Subscription<MessageT>::SharedPtr
+#define AUTOWARE_PUBLISHER_PTR(MessageT) typename agnocast::Publisher<MessageT>::SharedPtr
+
+#define AUTOWARE_POLLING_SUBSCRIBER(MessageT) typename agnocast::PollingSubscriber<MessageT>
+
+#define AUTOWARE_CREATE_SUBSCRIPTION(message_type, topic, qos, callback, options) \
+  agnocast::create_subscription<message_type>(this, topic, qos, callback, options)
+#define AUTOWARE_CREATE_PUBLISHER2(message_type, arg1, arg2) \
+  agnocast::create_publisher<message_type>(this, arg1, arg2)
+#define AUTOWARE_CREATE_PUBLISHER3(message_type, arg1, arg2, arg3) \
+  agnocast::create_publisher<message_type>(this, arg1, arg2, arg3)
+
+#define AUTOWARE_SUBSCRIPTION_OPTIONS agnocast::SubscriptionOptions
+#define AUTOWARE_PUBLISHER_OPTIONS agnocast::PublisherOptions
+
+#define ALLOCATE_OUTPUT_MESSAGE(publisher) publisher->borrow_loaned_message()
+
+#else
+
+#include "autoware/universe_utils/ros/polling_subscriber.hpp"
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <memory>
+
+#define AUTOWARE_MESSAGE_PTR(MessageT) std::shared_ptr<MessageT>
+#define AUTOWARE_SUBSCRIPTION_PTR(MessageT) typename rclcpp::Subscription<MessageT>::SharedPtr
+#define AUTOWARE_PUBLISHER_PTR(MessageT) typename rclcpp::Publisher<MessageT>::SharedPtr
+
+#define AUTOWARE_POLLING_SUBSCRIBER(MessageT) \
+  typename autoware::universe_utils::InterProcessPollingSubscriber<MessageT>
+
+#define AUTOWARE_CREATE_SUBSCRIPTION(message_type, topic, qos, callback, options) \
+  this->create_subscription<message_type>(topic, qos, callback, options)
+#define AUTOWARE_CREATE_PUBLISHER2(message_type, arg1, arg2) \
+  this->create_publisher<message_type>(arg1, arg2)
+#define AUTOWARE_CREATE_PUBLISHER3(message_type, arg1, arg2, arg3) \
+  this->create_publisher<message_type>(arg1, arg2, arg3)
+
+#define AUTOWARE_SUBSCRIPTION_OPTIONS rclcpp::SubscriptionOptions
+#define AUTOWARE_PUBLISHER_OPTIONS rclcpp::PublisherOptions
+
+#define ALLOCATE_OUTPUT_MESSAGE(publisher) \
+  std::make_unique<typename std::remove_reference<decltype(*publisher)>::type::ROSMessageType>()
```

Comment:
I recommend

```suggestion
  std::make_unique<typename std::decay_t<std::remove_pointer_t<decltype(publisher)>>::ROSMessageType>()
```

just in case.

And I strongly hope to there is a working example or test

Reply:
@soblin 
Is it ok to add tests in another pr?

Reply:
@sykwer Sure !